File: М11_ClusterVMS/clustervms-go/cluster/variables.go
// Package cluster is ClusterVMS — М11 — in Go: М10's platform shape across
// several servers. Built ON М10's vmsserver module (imported, not copied):
// the same psimplatform contract and the same vms controller, worker and
// archive resource. This package supplies what a cluster adds and nothing else:
//
//	variables.go    Nomad Variables over HTTP (ModifyIndex, cas) — and the fake with the promised semantics
//	objectstore.go  objects as Variables (this cluster's choice), HTTP, a directory, S3 (s3.go)
//	worker.go       the worker as an allocation: a slot from NOMAD_ALLOC_INDEX, labels from the server
//	controller.go   the controller as a job: placement under label constraints; the snapshot for М12
//	resource.go     the archive resource as a system job: the VMS's routes on the platform's server
//	timeline.go     one camera across two resources; *unavailable*, never *lost*
//	directory.go    where is camera 7 — one scan of vms/workers/*
//	console.go      the cluster console, standard library
//
// Standard library only: the Variables client is net/http against Nomad's
// API (github.com/hashicorp/nomad/api would do the same in more code and an
// MPL-2.0 dependency).
package cluster

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"sort"
	"strings"
	"sync"
	"time"

	p "vmsserver/psimplatform"
)

// The platform's names, re-exported so a cluster program imports one package.
type (
	Items     = p.Items
	Variables = p.Variables
)

var (
	ErrConflict  = p.ErrConflict
	ErrForbidden = p.ErrForbidden
)

const NoCAS = p.NoCAS

// NomadVariables speaks the HTTP API with the task's own workload-identity
// token (NOMAD_TOKEN).
type NomadVariables struct {
	Addr      string
	Token     string
	Namespace string
	Client    *http.Client
}

func NewNomadVariables() *NomadVariables {
	addr := os.Getenv("NOMAD_ADDR")
	if addr == "" {
		addr = "http://127.0.0.1:4646"
	}
	return &NomadVariables{Addr: strings.TrimRight(addr, "/"), Token: os.Getenv("NOMAD_TOKEN"),
		Namespace: "default", Client: &http.Client{Timeout: 5 * time.Second}}
}

type nomadVar struct {
	Path        string            `json:"Path"`
	Items       map[string]string `json:"Items"`
	ModifyIndex int64             `json:"ModifyIndex"`
}

func (n *NomadVariables) do(method, u string, body any) (int, []byte, error) {
	var rd io.Reader
	if body != nil {
		b, _ := json.Marshal(body)
		rd = bytes.NewReader(b)
	}
	req, err := http.NewRequest(method, u, rd)
	if err != nil {
		return 0, nil, err
	}
	req.Header.Set("X-Nomad-Token", n.Token)
	req.Header.Set("Content-Type", "application/json")
	resp, err := n.Client.Do(req)
	if err != nil {
		return 0, nil, err
	}
	defer resp.Body.Close()
	raw, _ := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	switch resp.StatusCode {
	case 409:
		return 409, raw, fmt.Errorf("%w: %.200s", ErrConflict, raw)
	case 403:
		return 403, raw, fmt.Errorf("%w: %s", ErrForbidden, u)
	case 404:
		return 404, nil, nil
	}
	if resp.StatusCode >= 300 {
		return resp.StatusCode, raw, fmt.Errorf("%s %s: HTTP %d: %.200s", method, u, resp.StatusCode, raw)
	}
	return resp.StatusCode, raw, nil
}

func (n *NomadVariables) Get(path string) (Items, int64, error) {
	status, raw, err := n.do("GET", fmt.Sprintf("%s/v1/var/%s?namespace=%s", n.Addr, path, n.Namespace), nil)
	if err != nil || status == 404 || len(raw) == 0 {
		return nil, 0, err
	}
	var v nomadVar
	if err := json.Unmarshal(raw, &v); err != nil {
		return nil, 0, err
	}
	return Items(v.Items), v.ModifyIndex, nil
}

func (n *NomadVariables) query(cas int64) string {
	q := "namespace=" + n.Namespace
	if cas != NoCAS {
		q += fmt.Sprintf("&cas=%d", cas)
	}
	return q
}

func (n *NomadVariables) Put(path string, items Items, cas int64) (int64, error) {
	_, raw, err := n.do("PUT", fmt.Sprintf("%s/v1/var/%s?%s", n.Addr, path, n.query(cas)), map[string]any{"Items": items})
	if err != nil {
		return 0, err
	}
	var v nomadVar
	if err := json.Unmarshal(raw, &v); err != nil {
		return 0, err
	}
	return v.ModifyIndex, nil
}

func (n *NomadVariables) List(prefix string) ([]string, error) {
	_, raw, err := n.do("GET", fmt.Sprintf("%s/v1/vars?prefix=%s&namespace=%s", n.Addr, url.QueryEscape(prefix), n.Namespace), nil)
	if err != nil {
		return nil, err
	}
	var vs []nomadVar
	if len(raw) > 0 {
		if err := json.Unmarshal(raw, &vs); err != nil {
			return nil, err
		}
	}
	out := make([]string, 0, len(vs))
	for _, v := range vs {
		out = append(out, v.Path)
	}
	sort.Strings(out)
	return out, nil
}

func (n *NomadVariables) Delete(path string, cas int64) error {
	_, _, err := n.do("DELETE", fmt.Sprintf("%s/v1/var/%s?%s", n.Addr, path, n.query(cas)), nil)
	return err
}

// FakeVariables is one raft log for the whole cluster, in memory, with the
// semantics the docs promise: a raft-assigned ModifyIndex, PUT ?cas=<index>
// succeeding only if the index still matches, a conflict otherwise. Optional
// ACL: a writer may only put under the prefixes it was granted. Nothing in
// the fake is Nomad; everything in it is what Nomad promises.
type FakeVariables struct {
	s      *fakeState
	Writer string // "who am I" for the ACL check; "" bypasses it
}

type fakeState struct {
	mu        sync.Mutex
	raftIndex int64
	items     map[string]fakeEntry
	acl       map[string][]string
}

type fakeEntry struct {
	items Items
	index int64
}

func NewFakeVariables() *FakeVariables {
	return &FakeVariables{s: &fakeState{raftIndex: 1000, items: map[string]fakeEntry{}, acl: map[string][]string{}}}
}

// AsWriter is the same raft seen through one identity — what a task's
// workload identity token is under a Nomad ACL policy. With prefixes given,
// they become that writer's grant.
func (f *FakeVariables) AsWriter(writer string, allowed ...string) *FakeVariables {
	f.s.mu.Lock()
	if allowed != nil {
		f.s.acl[writer] = allowed
	}
	f.s.mu.Unlock()
	return &FakeVariables{s: f.s, Writer: writer}
}

func (f *FakeVariables) acl(path string) error {
	if f.Writer == "" || len(f.s.acl) == 0 {
		return nil
	}
	if !p.Allowed(path, f.s.acl[f.Writer]) {
		return fmt.Errorf("%w: %s may not write %s", ErrForbidden, f.Writer, path)
	}
	return nil
}

func (f *FakeVariables) Get(path string) (Items, int64, error) {
	f.s.mu.Lock()
	defer f.s.mu.Unlock()
	e, ok := f.s.items[path]
	if !ok {
		return nil, 0, nil
	}
	out := make(Items, len(e.items))
	for k, v := range e.items {
		out[k] = v
	}
	return out, e.index, nil
}

func (f *FakeVariables) Put(path string, items Items, cas int64) (int64, error) {
	f.s.mu.Lock()
	defer f.s.mu.Unlock()
	if err := f.acl(path); err != nil {
		return 0, err
	}
	current := f.s.items[path].index
	if cas != NoCAS && cas != current {
		return 0, fmt.Errorf("%w: cas=%d but ModifyIndex=%d", ErrConflict, cas, current)
	}
	f.s.raftIndex++
	cp := make(Items, len(items))
	for k, v := range items {
		cp[k] = v
	}
	f.s.items[path] = fakeEntry{cp, f.s.raftIndex}
	return f.s.raftIndex, nil
}

func (f *FakeVariables) List(prefix string) ([]string, error) {
	f.s.mu.Lock()
	defer f.s.mu.Unlock()
	out := []string{}
	for pth := range f.s.items {
		if strings.HasPrefix(pth, prefix) {
			out = append(out, pth)
		}
	}
	sort.Strings(out)
	return out, nil
}

func (f *FakeVariables) Delete(path string, cas int64) error {
	f.s.mu.Lock()
	defer f.s.mu.Unlock()
	if err := f.acl(path); err != nil {
		return err
	}
	current := f.s.items[path].index
	if cas != NoCAS && cas != current {
		return fmt.Errorf("%w: cas=%d but ModifyIndex=%d", ErrConflict, cas, current)
	}
	delete(f.s.items, path)
	f.s.raftIndex++
	return nil
}

File: М11_ClusterVMS/clustervms/cluster/variables.py
"""Nomad Variables — the cluster's small, consistent store: М10's
`w2cplatform.variables.Variables` contract, implemented by raft.

`NomadVariables` speaks the HTTP API with the task's own workload-identity
token (NOMAD_TOKEN). `FakeVariables` is the same contract in memory, with
the semantics the docs promise: a raft-assigned ModifyIndex, PUT ?cas=<index>
succeeding only if the index still matches, 409 otherwise. Nothing in the
fake is Nomad; everything in it is what Nomad promises, and the tests run
against it in milliseconds.
"""
from __future__ import annotations

import json
import os
import threading
import urllib.error
import urllib.request
from urllib.parse import quote
from typing import Protocol


from w2cplatform.variables import Conflict, Forbidden   # noqa: E402  the platform's exceptions: one class, so a CAS retry catches ours too
from w2cplatform.limits import check
from w2cplatform.variables import items_bytes, register_scheme, safe_path


class Variables(Protocol):
    def get(self, path: str) -> tuple[dict | None, int]: ...
    def put(self, path: str, items: dict, cas: int | None = None) -> int: ...
    def list(self, prefix: str) -> list[str]: ...
    def delete(self, path: str, cas: int | None = None) -> None: ...


# `nomad://host:port` — the scheme this module answers for. Registered at the bottom of the file, so a
# process that imports `cluster` can say `CONFIG_URL=nomad://…` and the platform never names Nomad.
# `writer`/`acl` are ignored here: on a cluster the ACL is the task's own workload identity, enforced by
# the server, not by the client (which is the point of a real store).
def _open_nomad(url: str, writer: str | None = None, acl: dict | None = None) -> "NomadVariables":
    rest = url.split("://", 1)[1]
    return NomadVariables(addr=f"http://{rest}" if rest else None)


# What Nomad will hold in one Variable: every key and every value in it, together. The number is the
# scheduler's constant `maxVariableSize = 65536`, not ours — the request to make it configurable has been
# open since 2022, answered with "we don't want to give users a new way to break their clusters". This is
# the line that turns it from prose into something the platform can read (М10A Lesson 26).
NOMAD_MAX_VARIABLE_BYTES = 65536


class NomadVariables:
    max_bytes = NOMAD_MAX_VARIABLE_BYTES

    def __init__(self, addr: str | None = None, token: str | None = None, namespace: str = "default",
                 timeout: float = 5.0):
        self.addr = (addr or os.environ.get("NOMAD_ADDR", "http://127.0.0.1:4646")).rstrip("/")
        self.token = token or os.environ.get("NOMAD_TOKEN", "")
        self.namespace = namespace
        self.timeout = timeout

    def _req(self, method: str, url: str, body: dict | None = None):
        data = json.dumps(body).encode() if body is not None else None
        req = urllib.request.Request(url, data=data, method=method)
        req.add_header("X-Nomad-Token", self.token)
        req.add_header("Content-Type", "application/json")
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as r:
                raw = r.read()
                return r.status, (json.loads(raw) if raw else None)
        except urllib.error.HTTPError as e:
            if e.code == 409:
                raise Conflict(e.read().decode(errors="replace")[:200]) from None
            if e.code == 403:
                raise Forbidden(url) from None
            if e.code == 404:
                return 404, None
            raise

    # The key, checked and then percent-encoded for a URL. `safe_path` is the platform's rule and it
    # matters MORE here than on a box: this path goes into a URL, where a `..` meets HTTP normalisation —
    # ours, a proxy's, or the server's — and the ACL that let it through was matched against the string
    # BEFORE that. `quote` covers the rest: a space or a `#` in a unit's name would otherwise end the path.
    def _key(self, path: str) -> str:
        return quote(safe_path(path), safe="/")

    def get(self, path: str) -> tuple[dict | None, int]:
        status, body = self._req("GET", f"{self.addr}/v1/var/{self._key(path)}?namespace={self.namespace}")
        if status == 404 or body is None:
            return None, 0
        return dict(body["Items"]), int(body["ModifyIndex"])

    def put(self, path: str, items: dict, cas: int | None = None) -> int:
        # Refused here, by the number this store declares, rather than by a 400 from the server after the
        # round trip: the caller gets the size and the limit, and nothing was sent.
        check(path, items_bytes(items), self.max_bytes)
        q = f"namespace={self.namespace}" + (f"&cas={cas}" if cas is not None else "")
        _, body = self._req("PUT", f"{self.addr}/v1/var/{self._key(path)}?{q}", {"Items": {k: str(v) for k, v in items.items()}})
        return int(body["ModifyIndex"])

    def list(self, prefix: str) -> list[str]:
        status, body = self._req("GET", f"{self.addr}/v1/vars?prefix={quote(prefix, safe=chr(47))}&namespace={self.namespace}")
        return [v["Path"] for v in (body or [])]

    def delete(self, path: str, cas: int | None = None) -> None:
        q = f"namespace={self.namespace}" + (f"&cas={cas}" if cas is not None else "")
        self._req("DELETE", f"{self.addr}/v1/var/{self._key(path)}?{q}")


class FakeVariables:
    """One raft log for the whole cluster, in memory. Optional ACL: a writer
    id may only put under the prefixes it was granted."""

    def __init__(self):
        self._lock = threading.Lock()
        self._raft_index = 1000
        self._items: dict[str, tuple[dict, int]] = {}
        self.acl: dict[str, list[str]] = {}        # writer -> allowed prefixes
        self.writer: str | None = None            # "who am I" for the ACL check

    def as_writer(self, writer: str, allowed: list[str] | None = None) -> "FakeVariables":
        """The same raft seen through one identity — what a task's workload
        identity token is under a Nomad ACL policy."""
        v = FakeVariables.__new__(FakeVariables)
        v.__dict__ = self.__dict__.copy(); v.writer = writer
        if allowed is not None:
            self.acl[writer] = list(allowed)
        return v

    def _acl(self, path):
        if self.writer is not None and self.acl:
            allowed = self.acl.get(self.writer, [])
            if not any(path == p or (p.endswith("*") and path.startswith(p[:-1])) for p in allowed):
                raise Forbidden(f"{self.writer} may not write {path}")

    def get(self, path):
        safe_path(path)
        with self._lock:
            if path not in self._items:
                return None, 0
            items, idx = self._items[path]
            return dict(items), idx

    def put(self, path, items, cas=None):
        safe_path(path)          # the same key rule as the real store: the fake may not be laxer
        self._acl(path)
        with self._lock:
            _, current = self._items.get(path, (None, 0))
            if cas is not None and cas != current:
                raise Conflict(f"cas={cas} but ModifyIndex={current}")
            self._raft_index += 1
            self._items[path] = ({k: str(v) for k, v in items.items()}, self._raft_index)
            return self._raft_index

    def list(self, prefix):
        with self._lock:
            return sorted(p for p in self._items if p.startswith(prefix))

    def delete(self, path, cas=None):
        safe_path(path)
        self._acl(path)
        with self._lock:
            _, current = self._items.get(path, (None, 0))
            if cas is not None and cas != current:
                raise Conflict(f"cas={cas} but ModifyIndex={current}")
            self._items.pop(path, None)
            self._raft_index += 1


register_scheme("nomad", _open_nomad)
